srml/babe/src/lib.rs

```diff
@@ -216,42 +262,53 @@ impl<T: Trait> OnTimestampSet<T::Moment> for Module<T> {
 	fn on_timestamp_set(_moment: T::Moment) { }
 }
 
-impl<T: Trait> session::OneSessionHandler<T::AccountId> for Module<T> {
-	type Key = AuthorityId;
+impl<T: Trait> session::ShouldEndSession<u64> for Module<T> {
+	fn should_end_session(_now: u64) -> bool {
+		LastSlotInEpoch::take()
+	}
+}
 
-	fn on_new_session<'a, I: 'a>(changed: bool, validators: I)
+impl<T: Trait + staking::Trait> session::OneSessionHandler<T::AccountId> for Module<T> {
```

Comment:
Not sure if this coupling with staking module is the best approach, but I'm not very familiar with it.

Reply:
It was the only way I could figure out for getting weights into BABE.

Reply:
You mean `CurrencyToVote`? You could just add this to your trait as well.